src/disk.hpp

```diff
@@ -232,14 +232,23 @@ struct BufferedDisk : Disk
         NeedReadCache();
         // all allocations need 7 bytes head-room, since
         // SliceInt64FromBytes() may overrun by 7 bytes
-        if (read_buffer_start_ <= begin && read_buffer_start_ + read_buffer_size_ > begin + length + 7) {
+        if (read_buffer_start_ <= begin
+            && read_buffer_start_ + read_buffer_size_ >= begin + length
+            && read_buffer_start_ + read_ahead >= begin + length + 7)
+        {
             // if the read is entirely inside the buffer, just return it
             return read_buffer_.get() + (begin - read_buffer_start_);
         }
-        else if (begin >= read_buffer_start_ || begin == 0) {
+        else if (begin >= read_buffer_start_ || begin == 0 || read_buffer_start_ == std::uint64_t(-1)) {
```

Comment:
the `begin == 0` condition here was meant to catch the initial read, since we perform forward sequential scans. However, with the `FilteredDisk` (where we save a read/write pass to compact the table) our first read may not actually be at 0, but a few entries in (if the first ones were dropped).

This fixes the condition to also say, "if we don't have a cache yet" (i.e. `read_buffer_start == std::uint64_t(-1)`)

Reply:
if there truly is a performance regression with this patch (which seems unlikely to me), I would think it's the fact that there's one more condition here.